### PR TITLE
fix: multitenancy recipe

### PR DESCRIPTION
**Pull request by sattvikc**
## Summary of change

(A few sentences about this PR)

## Related issues

-   Link to issue1 here
-   Link to issue1 here

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots and videos!)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates

-   [ ] Changelog has been updated
-   [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `lib/ts/version.ts`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [ ] Changes to the version if needed
    -   In `package.json`
    -   In `package-lock.json`
    -   In `lib/ts/version.ts`
-   [ ] Had run `npm run build-pretty`
-   [ ] Had installed and ran the pre-commit hook
-   [ ] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If have added a new web framework, update the `add-ts-no-check.js` file to include that
-   [ ] If added a new recipe / api interface, then make sure that the implementation of it uses NON arrow functions only (like `someFunc: function () {..}`).
-   [ ] If added a new recipe, then make sure to expose it inside the recipe folder present in the root of this repo. We also need to expose its types.

## Remaining TODOs for this PR

-   [ ] Item1
-   [ ] Item2

### CODE REVIEW

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/recipe/multitenancy/index.ts [hunk 1]
```diff
@@ -43,7 +43,7 @@ export default class Wrapper {
         userContext?: any
     ): Promise<{
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply lib/ts/recipe/multitenancy/index.ts:49]: Done

File: lib/ts/recipe/multitenancy/index.ts [hunk 2]
```diff
@@ -43,7 +43,7 @@ export default class Wrapper {
...
         status: "OK";
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply lib/ts/recipe/multitenancy/index.ts:50]: Done

File: lib/ts/version.ts [hunk 3]
```diff
@@ -14,7 +14,7 @@
  */
 export const version = "14.1.2";
 
-export const cdiSupported = ["2.21"];
+export const cdiSupported = ["3.0", "2.21"];
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply lib/ts/version.ts:17]: Done

File: lib/ts/recipe/multitenancy/recipeImplementation.ts [hunk 4]
```diff
@@ -1,57 +1,110 @@
...
+                    thirdPartyId,
+                }
+            );
+            return response;
         },
 
-        listThirdPartyConfigsForThirdPartyId: async function () {
-            // TODO
-            throw new Error("Not implemented");
+        listThirdPartyConfigsForThirdPartyId: async function ({ thirdPartyId }) {
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply lib/ts/recipe/multitenancy/recipeImplementation.ts]: Done

File: lib/ts/recipe/multitenancy/recipeImplementation.ts [hunk 5]
```diff
@@ -1,57 +1,110 @@
...
+import NormalisedURLPath from "../../normalisedURLPath";
+import { DEFAULT_TENANT_ID } from "./constants";
+import { ProviderConfig } from "../thirdparty/types";
 
-export default function getRecipeInterface(_: Querier): RecipeInterface {
+export default function getRecipeInterface(querier: Querier): RecipeInterface {
     return {
         getTenantId: async function ({ tenantIdFromFrontend }) {
-            // TODO
+            // TODO do we need this function?
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply lib/ts/recipe/multitenancy/recipeImplementation.ts]: Done

File: lib/ts/recipe/multitenancy/recipeImplementation.ts [hunk 6]
```diff
@@ -1,57 +1,110 @@
 import { RecipeInterface } from "./";
 import { Querier } from "../../querier";
+import NormalisedURLPath from "../../normalisedURLPath";
+import { DEFAULT_TENANT_ID } from "./constants";
+import { ProviderConfig } from "../thirdparty/types";
 
-export default function getRecipeInterface(_: Querier): RecipeInterface {
+export default function getRecipeInterface(querier: Querier): RecipeInterface {
     return {
         getTenantId: async function ({ tenantIdFromFrontend }) {
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply lib/ts/recipe/multitenancy/recipeImplementation.ts:8]: introduces complexity with usercontext

File: lib/ts/recipe/multitenancy/types.ts [hunk 7]
```diff
@@ -138,15 +140,23 @@ export type RecipeInterface = {
...
+        userId: string;
         userContext: any;
     }) => Promise<{
         status: "OK";
-        tenants: {
-            tenantId: string;
-            providers: ProviderConfig[];
-        }[];
+        wasAlreadyAssociated: boolean;
+    }>;
```
Comment (rishabhpoddar) [inline lib/ts/recipe/multitenancy/types.ts:152]: add more status codes cause core returns those status codes

File: lib/ts/recipe/multitenancy/types.ts [hunk 8]
```diff
@@ -84,6 +84,7 @@ export type RecipeInterface = {
             emailPasswordEnabled?: boolean;
             passwordlessEnabled?: boolean;
             thirdPartyEnabled?: boolean;
+            coreConfig?: any;
```
Comment (rishabhpoddar) [inline lib/ts/recipe/multitenancy/types.ts:87]:
```suggestion
            coreConfig?: {[key: string]: any};
```

File: lib/ts/recipe/multitenancy/types.ts [hunk 9]
```diff
@@ -112,6 +113,7 @@ export type RecipeInterface = {
             enabled: boolean;
             providers: ProviderConfig[];
         };
+        coreConfig: any;
```
Comment (rishabhpoddar) [inline lib/ts/recipe/multitenancy/types.ts:116]:
```suggestion
        coreConfig: {[key: string]: any};;
```

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/recipe/multitenancy/types.ts (hunk 7 above)
Comment (sattvikc) [reply lib/ts/recipe/multitenancy/types.ts]: Done

File: lib/ts/recipe/multitenancy/types.ts (hunk 8 above)
Comment (sattvikc) [reply lib/ts/recipe/multitenancy/types.ts]: Done

File: lib/ts/recipe/multitenancy/types.ts (hunk 9 above)
Comment (sattvikc) [reply lib/ts/recipe/multitenancy/types.ts]: Done

**Review by rishabhpoddar — APPROVED** [review]

**Pull request merged**